### PR TITLE
CHE-476 Add support for git urls with credentials

### PR DESCRIPTION
**Pull request by vinokurig**
### What does this PR do?

Adds possibility to perform remote git operations by http or https url with credentials in the url.
It is needed for example to clone from private repository with http or https url that contains username and password.
### What issues does this PR fix or reference?

https://github.com/codenvy/codenvy/issues/476
### Tests written?

Yes

@sleshchenko @skabashnyuk please review

### CODE REVIEW

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 1]
```diff
@@ -1589,18 +1593,19 @@ protected JSch getJSch(final OpenSshConfig.Host hc, FS fs) throws JSchException
...
+                command.setTransportConfigCallback(transport -> {
+                    SshTransport sshTransport = (SshTransport)transport;
+                    sshTransport.setSshSessionFactory(sshSessionFactory);
                 });
             } else {
                 credentials = credentialsLoader.getUserCredential(remoteUrl);
                 if (credentials != null) {
                     command.setCredentialsProvider(new UsernamePasswordCredentialsProvider(credentials.getUserName(),
                                                                                            credentials.getPassword()));
+                } else if (remoteUrl != null && GIT_URL_WITH_CREDENTIALS_PATTERN.matcher(remoteUrl).matches()) {
```
Comment (sleshchenko) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: I think credentials in url should have higher priority than credentials provided by credentials loader

File: wsagent/che-core-git-impl-jgit/src/test/java/org/eclipse/che/git/impl/jgit/JGitConnectionTest.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    JGitConnection jGitConnection;
+
+    @DataProvider(name = "gitUrlsWithCredentialsProvider")
+    public static Object[][] gitUrlsWithCredentials() {
+        return new Object[][]{{"http://username:password@host.xz/path/to/repo.git"},
+                              {"https://username:password@host.xz/path/to/repo.git"}};
+    }
+
+    @Test(dataProvider = "gitUrlsWithCredentials")
+    public void shouldExecuteRemoteCommandByHttpOrHttpsUrlWithCredentials(String url) throws Exception {
```
Comment (sleshchenko) [inline wsagent/che-core-git-impl-jgit/src/test/java/org/eclipse/che/git/impl/jgit/JGitConnectionTest.java:74]: Please add test for testing case when url doesn't contain credentials

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java (hunk 1 above)
Comment (vinokurig) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: done

File: wsagent/che-core-git-impl-jgit/src/test/java/org/eclipse/che/git/impl/jgit/JGitConnectionTest.java (hunk 2 above)
Comment (vinokurig) [inline wsagent/che-core-git-impl-jgit/src/test/java/org/eclipse/che/git/impl/jgit/JGitConnectionTest.java:74]: done

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 3]
```diff
@@ -1610,18 +1612,21 @@ protected JSch getJSch(final OpenSshConfig.Host hc, FS fs) throws JSchException
...
                 };
-                command.setTransportConfigCallback(new TransportConfigCallback() {
-                    @Override
-                    public void configure(Transport transport) {
-                        SshTransport sshTransport = (SshTransport)transport;
-                        sshTransport.setSshSessionFactory(sshSessionFactory);
-                    }
+                command.setTransportConfigCallback(transport -> {
+                    SshTransport sshTransport = (SshTransport)transport;
+                    sshTransport.setSshSessionFactory(sshSessionFactory);
```
Comment (sleshchenko) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]:
```
sshTransport.setSshSessionFactory((SshTransport)transport);
```
?

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java [hunk 4]
```diff
@@ -1610,18 +1612,21 @@ protected JSch getJSch(final OpenSshConfig.Host hc, FS fs) throws JSchException
...
             } else {
-                credentials = credentialsLoader.getUserCredential(remoteUrl);
-                if (credentials != null) {
-                    command.setCredentialsProvider(new UsernamePasswordCredentialsProvider(credentials.getUserName(),
-                                                                                           credentials.getPassword()));
+                if (remoteUrl != null && GIT_URL_WITH_CREDENTIALS_PATTERN.matcher(remoteUrl).matches()) {
+                    String username = remoteUrl.substring(remoteUrl.indexOf("://") + 3, remoteUrl.lastIndexOf(":"));
+                    String password = remoteUrl.substring(remoteUrl.lastIndexOf(":") + 1, remoteUrl.indexOf("@"));
+                    command.setCredentialsProvider(new UsernamePasswordCredentialsProvider(username, password));
+                } else{
```
Comment (sleshchenko) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: add space after `else`

**Comment by sleshchenko** [comment]
LGTM

File: wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java (hunk 4 above)
Comment (vinokurig) [inline wsagent/che-core-git-impl-jgit/src/main/java/org/eclipse/che/git/impl/jgit/JGitConnection.java]: done

**Comment by skabashnyuk** [comment]
ок

**Pull request merged**